File: backend/src/schemas/expense.py
"""Expense request/response schemas.

Implements T127 from tasks.md.
"""
from datetime import date
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from ..models.expense import ExpenseCategory, ExpenseStatus


class ExpenseCreateRequest(BaseModel):
    """Request schema for creating a new expense."""

    property_id: UUID = Field(..., description="Property ID where expense was incurred")
    amount: Decimal = Field(..., gt=0, description="Expense amount")
    currency: str = Field(default="NPR", max_length=3, description="Currency code")
    category: ExpenseCategory = Field(..., description="Expense category")
    description: str = Field(..., min_length=1, max_length=2000, description="Expense description")
    vendor_name: Optional[str] = Field(None, max_length=255, description="Vendor/service provider name")
    invoice_number: Optional[str] = Field(None, max_length=100, description="Invoice or receipt number")
    paid_by: str = Field(default="intermediary", max_length=100, description="Who paid for this expense")
    is_reimbursable: bool = Field(default=True, description="Whether expense is reimbursable")
    expense_date: date = Field(..., description="Date when expense was incurred")
    notes: Optional[str] = Field(None, max_length=2000, description="Additional notes")

    @field_validator('expense_date')
    @classmethod
    def validate_expense_date(cls, v: date) -> date:
        """Validate expense date is not in the future."""
        if v > date.today():
            raise ValueError("Expense date cannot be in the future")
        return v

    @field_validator('amount')
    @classmethod
    def validate_amount(cls, v: Decimal) -> Decimal:
        """Validate amount has max 2 decimal places."""
        if v.as_tuple().exponent < -2:
            raise ValueError("Amount cannot have more than 2 decimal places")
        return v

    model_config = {
        "json_schema_extra": {
            "example": {
                "property_id": "123e4567-e89b-12d3-a456-426614174000",
                "amount": 5000.00,
                "currency": "NPR",
                "category": "maintenance",
                "description": "Plumbing repair for leaking pipe in bathroom",
                "vendor_name": "ABC Plumbing Services",
                "invoice_number": "INV-2024-001",
                "paid_by": "intermediary",
                "is_reimbursable": True,
                "expense_date": "2024-01-15",
                "notes": "Emergency repair required",
            }
        }
    }


class ExpenseUpdateRequest(BaseModel):
    """Request schema for updating an expense."""

    amount: Optional[Decimal] = Field(None, gt=0, description="Expense amount")
    category: Optional[ExpenseCategory] = Field(None, description="Expense category")
    description: Optional[str] = Field(None, min_length=1, max_length=2000)
    vendor_name: Optional[str] = Field(None, max_length=255)
    invoice_number: Optional[str] = Field(None, max_length=100)
    expense_date: Optional[date] = Field(None, description="Date when expense was incurred")
    notes: Optional[str] = Field(None, max_length=2000)

    @field_validator('expense_date')
    @classmethod
    def validate_expense_date(cls, v: Optional[date]) -> Optional[date]:
        """Validate expense date is not in the future."""
        if v and v > date.today():
            raise ValueError("Expense date cannot be in the future")
        return v


class ExpenseApprovalRequest(BaseModel):
    """Request schema for approving/rejecting an expense."""

    status: ExpenseStatus = Field(..., description="Approval status (approved/rejected)")
    rejection_reason: Optional[str] = Field(None, max_length=1000, description="Reason if rejected")

    @field_validator('status')
    @classmethod
    def validate_status(cls, v: ExpenseStatus) -> ExpenseStatus:
        """Validate status is either approved or rejected."""
        if v not in [ExpenseStatus.APPROVED, ExpenseStatus.REJECTED]:
            raise ValueError("Status must be either 'approved' or 'rejected'")
        return v

    @field_validator('rejection_reason')
    @classmethod
    def validate_rejection_reason(cls, v: Optional[str], info) -> Optional[str]:
        """Require rejection reason if status is rejected."""
        status = info.data.get('status')
        if status == ExpenseStatus.REJECTED and not v:
            raise ValueError("Rejection reason is required when rejecting an expense")
        return v


class ExpenseReimbursementRequest(BaseModel):
    """Request schema for marking expense as reimbursed."""

    reimbursed_date: date = Field(..., description="Date when reimbursement was made")

    @field_validator('reimbursed_date')
    @classmethod
    def validate_reimbursed_date(cls, v: date) -> date:
        """Validate reimbursed date is not in the future."""
        if v > date.today():
            raise ValueError("Reimbursed date cannot be in the future")
        return v


class ExpenseResponse(BaseModel):
    """Response schema for expense."""

    id: UUID
    property_id: UUID
    recorded_by: UUID
    approved_by: Optional[UUID]
    amount: Decimal
    currency: str
    category: ExpenseCategory
    status: ExpenseStatus
    description: str
    vendor_name: Optional[str]
    invoice_number: Optional[str]
    receipt_url: Optional[str]
    paid_by: str
    is_reimbursable: bool
    is_reimbursed: bool
    reimbursed_date: Optional[date]
    expense_date: date
    approved_date: Optional[date]
    created_at: str
    updated_at: str
    notes: Optional[str]
    rejection_reason: Optional[str]

    model_config = {"from_attributes": True}


class ExpenseListResponse(BaseModel):
    """Response schema for expense list with pagination."""

    total: int = Field(..., description="Total number of expenses")
    page: int = Field(..., description="Current page number")
    page_size: int = Field(..., description="Number of items per page")
    expenses: list[ExpenseResponse] = Field(..., description="List of expenses")


class ExpenseSummary(BaseModel):
    """Summary statistics for expenses."""

    total_expenses: int = Field(..., description="Total number of expenses")
    total_amount: Decimal = Field(..., description="Total expense amount")
    pending_amount: Decimal = Field(..., description="Amount pending approval")
    approved_amount: Decimal = Field(..., description="Amount approved")
    reimbursed_amount: Decimal = Field(..., description="Amount reimbursed")
    outstanding_amount: Decimal = Field(..., description="Amount approved but not reimbursed")
    by_category: dict[str, Decimal] = Field(..., description="Amount by category")


__all__ = [
    "ExpenseCreateRequest",
    "ExpenseUpdateRequest",
    "ExpenseApprovalRequest",
    "ExpenseReimbursementRequest",
    "ExpenseResponse",
    "ExpenseListResponse",
    "ExpenseSummary",
]

File: mobile/lib/models/expense.dart
/// Expense model for local storage and API communication.
///
/// Implements T134 from tasks.md.
library;

import 'package:flutter/foundation.dart';

/// Expense category enum matching backend ExpenseCategory
enum ExpenseCategory {
  maintenance,
  repair,
  cleaning,
  landscaping,
  security,
  utilities,
  insurance,
  taxes,
  legal,
  administrative,
  other;

  String toJson() => name;

  static ExpenseCategory fromJson(String json) {
    switch (json.toLowerCase()) {
      case 'maintenance':
        return ExpenseCategory.maintenance;
      case 'repair':
        return ExpenseCategory.repair;
      case 'cleaning':
        return ExpenseCategory.cleaning;
      case 'landscaping':
        return ExpenseCategory.landscaping;
      case 'security':
        return ExpenseCategory.security;
      case 'utilities':
        return ExpenseCategory.utilities;
      case 'insurance':
        return ExpenseCategory.insurance;
      case 'taxes':
        return ExpenseCategory.taxes;
      case 'legal':
        return ExpenseCategory.legal;
      case 'administrative':
        return ExpenseCategory.administrative;
      case 'other':
        return ExpenseCategory.other;
      default:
        throw ArgumentError('Invalid expense category: $json');
    }
  }

  String get displayName {
    switch (this) {
      case ExpenseCategory.maintenance:
        return 'Maintenance';
      case ExpenseCategory.repair:
        return 'Repair';
      case ExpenseCategory.cleaning:
        return 'Cleaning';
      case ExpenseCategory.landscaping:
        return 'Landscaping';
      case ExpenseCategory.security:
        return 'Security';
      case ExpenseCategory.utilities:
        return 'Utilities';
      case ExpenseCategory.insurance:
        return 'Insurance';
      case ExpenseCategory.taxes:
        return 'Taxes';
      case ExpenseCategory.legal:
        return 'Legal';
      case ExpenseCategory.administrative:
        return 'Administrative';
      case ExpenseCategory.other:
        return 'Other';
    }
  }
}

/// Expense status enum matching backend ExpenseStatus
enum ExpenseStatus {
  pending,
  approved,
  rejected,
  reimbursed;

  String toJson() => name;

  static ExpenseStatus fromJson(String json) {
    switch (json.toLowerCase()) {
      case 'pending':
        return ExpenseStatus.pending;
      case 'approved':
        return ExpenseStatus.approved;
      case 'rejected':
        return ExpenseStatus.rejected;
      case 'reimbursed':
        return ExpenseStatus.reimbursed;
      default:
        throw ArgumentError('Invalid expense status: $json');
    }
  }

  String get displayName {
    switch (this) {
      case ExpenseStatus.pending:
        return 'Pending';
      case ExpenseStatus.approved:
        return 'Approved';
      case ExpenseStatus.rejected:
        return 'Rejected';
      case ExpenseStatus.reimbursed:
        return 'Reimbursed';
    }
  }
}

/// Expense model
@immutable
class Expense {
  const Expense({
    required this.id,
    required this.propertyId,
    required this.amount,
    required this.category,
    required this.expenseDate,
    required this.description,
    this.vendorName,
    this.invoiceNumber,
    this.receiptUrl,
    this.paidBy,
    required this.isReimbursable,
    required this.isReimbursed,
    this.reimbursedDate,
    required this.status,
    required this.recordedBy,
    this.approvedBy,
    this.approvedDate,
    this.rejectionReason,
    required this.createdAt,
    required this.updatedAt,
  });

  /// Create from JSON
  factory Expense.fromJson(Map<String, dynamic> json) {
    return Expense(
      id: json['id'] as String,
      propertyId: json['property_id'] as String,
      amount: (json['amount'] as num).toDouble(),
      category: ExpenseCategory.fromJson(json['category'] as String),
      expenseDate: DateTime.parse(json['expense_date'] as String),
      description: json['description'] as String,
      vendorName: json['vendor_name'] as String?,
      invoiceNumber: json['invoice_number'] as String?,
      receiptUrl: json['receipt_url'] as String?,
      paidBy: json['paid_by'] as String?,
      isReimbursable: json['is_reimbursable'] as bool,
      isReimbursed: json['is_reimbursed'] as bool,
      reimbursedDate: json['reimbursed_date'] != null
          ? DateTime.parse(json['reimbursed_date'] as String)
          : null,
      status: ExpenseStatus.fromJson(json['status'] as String),
      recordedBy: json['recorded_by'] as String,
      approvedBy: json['approved_by'] as String?,
      approvedDate: json['approved_date'] != null
          ? DateTime.parse(json['approved_date'] as String)
          : null,
      rejectionReason: json['rejection_reason'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  /// Unique expense identifier
  final String id;

  /// Property ID
  final String propertyId;

  /// Expense amount
  final double amount;

  /// Expense category
  final ExpenseCategory category;

  /// Expense date
  final DateTime expenseDate;

  /// Description
  final String description;

  /// Vendor name
  final String? vendorName;

  /// Invoice number
  final String? invoiceNumber;

  /// Receipt URL
  final String? receiptUrl;

  /// Person who paid
  final String? paidBy;

  /// Is reimbursable
  final bool isReimbursable;

  /// Is reimbursed
  final bool isReimbursed;

  /// Reimbursed date
  final DateTime? reimbursedDate;

  /// Expense status
  final ExpenseStatus status;

  /// User who recorded the expense
  final String recordedBy;

  /// User who approved the expense
  final String? approvedBy;

  /// Approved date
  final DateTime? approvedDate;

  /// Rejection reason
  final String? rejectionReason;

  /// Created timestamp
  final DateTime createdAt;

  /// Updated timestamp
  final DateTime updatedAt;

  /// Convert to JSON
  Map<String, dynamic> toJson() => {
        'id': id,
        'property_id': propertyId,
        'amount': amount,
        'category': category.toJson(),
        'expense_date': expenseDate.toIso8601String().split('T')[0],
        'description': description,
        'vendor_name': vendorName,
        'invoice_number': invoiceNumber,
        'receipt_url': receiptUrl,
        'paid_by': paidBy,
        'is_reimbursable': isReimbursable,
        'is_reimbursed': isReimbursed,
        'reimbursed_date': reimbursedDate?.toIso8601String().split('T')[0],
        'status': status.toJson(),
        'recorded_by': recordedBy,
        'approved_by': approvedBy,
        'approved_date': approvedDate?.toIso8601String(),
        'rejection_reason': rejectionReason,
        'created_at': createdAt.toIso8601String(),
        'updated_at': updatedAt.toIso8601String(),
      };

  /// Copy with
  Expense copyWith({
    String? id,
    String? propertyId,
    double? amount,
    ExpenseCategory? category,
    DateTime? expenseDate,
    String? description,
    String? vendorName,
    String? invoiceNumber,
    String? receiptUrl,
    String? paidBy,
    bool? isReimbursable,
    bool? isReimbursed,
    DateTime? reimbursedDate,
    ExpenseStatus? status,
    String? recordedBy,
    String? approvedBy,
    DateTime? approvedDate,
    String? rejectionReason,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) =>
      Expense(
        id: id ?? this.id,
        propertyId: propertyId ?? this.propertyId,
        amount: amount ?? this.amount,
        category: category ?? this.category,
        expenseDate: expenseDate ?? this.expenseDate,
        description: description ?? this.description,
        vendorName: vendorName ?? this.vendorName,
        invoiceNumber: invoiceNumber ?? this.invoiceNumber,
        receiptUrl: receiptUrl ?? this.receiptUrl,
        paidBy: paidBy ?? this.paidBy,
        isReimbursable: isReimbursable ?? this.isReimbursable,
        isReimbursed: isReimbursed ?? this.isReimbursed,
        reimbursedDate: reimbursedDate ?? this.reimbursedDate,
        status: status ?? this.status,
        recordedBy: recordedBy ?? this.recordedBy,
        approvedBy: approvedBy ?? this.approvedBy,
        approvedDate: approvedDate ?? this.approvedDate,
        rejectionReason: rejectionReason ?? this.rejectionReason,
        createdAt: createdAt ?? this.createdAt,
        updatedAt: updatedAt ?? this.updatedAt,
      );

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Expense && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;

  @override
  String toString() =>
      'Expense(id: $id, amount: $amount, category: ${category.displayName}, status: ${status.displayName})';
}

/// Expense summary model
@immutable
class ExpenseSummary {
  const ExpenseSummary({
    required this.totalAmount,
    required this.pendingAmount,
    required this.approvedAmount,
    required this.reimbursedAmount,
    required this.outstandingAmount,
    required this.byCategory,
  });

  /// Create from JSON
  factory ExpenseSummary.fromJson(Map<String, dynamic> json) {
    return ExpenseSummary(
      totalAmount: (json['total_amount'] as num).toDouble(),
      pendingAmount: (json['pending_amount'] as num).toDouble(),
      approvedAmount: (json['approved_amount'] as num).toDouble(),
      reimbursedAmount: (json['reimbursed_amount'] as num).toDouble(),
      outstandingAmount: (json['outstanding_amount'] as num).toDouble(),
      byCategory: (json['by_category'] as Map<String, dynamic>).map(
        (key, value) => MapEntry(key, (value as num).toDouble()),
      ),
    );
  }

  /// Total amount
  final double totalAmount;

  /// Pending amount
  final double pendingAmount;

  /// Approved amount
  final double approvedAmount;

  /// Reimbursed amount
  final double reimbursedAmount;

  /// Outstanding amount (approved but not reimbursed)
  final double outstandingAmount;

  /// Breakdown by category
  final Map<String, double> byCategory;

  /// Convert to JSON
  Map<String, dynamic> toJson() => {
        'total_amount': totalAmount,
        'pending_amount': pendingAmount,
        'approved_amount': approvedAmount,
        'reimbursed_amount': reimbursedAmount,
        'outstanding_amount': outstandingAmount,
        'by_category': byCategory,
      };
}
